draft-how-does-mutant-work.markdown

```diff
@@ -0,0 +1,431 @@
+
+Let's check out `Mutant::CLI.run`:
+
+```Ruby
+module Mutant
+  class CLI
+    include Adamantium::Flat, Equalizer.new(:config), Procto.call(:config)
+
+    def self.run(arguments)
+      Runner.call(Env::Bootstrap.call(call(arguments))).success?
+    rescue Error => exception
+      $stderr.puts(exception.message)
+      false
+    end
+  end
+end
+```
+
+The `include` line is what we need to understand first:
+
+```Ruby
+include Adamantium::Flat, Equalizer.new(:config), Procto.call(:config)
+```
+
+* `Adamantium::Flat`: This makes instances of this class immutable using the [Adamantium gem](https://github.com/dkubb/adamantium) (for a low level introduction check out [this blog post](https://troessner.svbtle.com/lessons-learned-from-some-of-the-best-ruby-codebases-out-there-part-3)))
+* `Equalizer.new(:config)`: This defines equality, equivalency and hash methods automatically for whatever `config` returns and is based on the [Equalizer](https://github.com/dkubb/equalizer) gem
+* `Procto.call(:config)`: [Procto](https://github.com/snusnu/procto) turns your Ruby object into a method object. Please check out [this blog post](https://troessner.svbtle.com/lessons-learned-from-some-of-the-best-ruby-codebases-out-there-part-3)) to read up on it.
+
+It is important that you memorize this, at least the `Procto` pattern because `Mutant` makes heavy use of those gems in almost every module.
+
+Now to `run`:
+
+```Ruby
+module Mutant
+  class CLI
+    def self.run(arguments)
+      Runner.call(Env::Bootstrap.call(call(arguments))).success?
+    rescue Error => exception
+      $stderr.puts(exception.message)
+      false
+    end
+  end
+end
+```
+
+Narrowing it down further this is the line we're interested in:
+
+```Ruby
+Runner.call(Env::Bootstrap.call(call(arguments))).success?
+```
+
+That line does pack quite a punch, doesn't it?
+
+Let me reformat this line to make clearer what's happening here:
+
+```Ruby
+Runner.call(
+  Env::Bootstrap.call(
+    call(arguments)
+  )
+).success?
+```
+
+Starting from the inside
+
+```Ruby
+  call(arguments)
+```
+
+translates to:
+
+```Ruby
+CLI.new(arguments).config
+```
+
+This is the part that `Procto` provides.
+
+So:
+
+* we're getting back a config based on our CLI arguments
+* then pass that to `Env::Bootstrap.call`
+* pass whatever comes back from that to our Runner
+
+**This leads us to 2 questions:**
+
+1.) What is the `config` and what is `Env::Bootstrap`?
+
+2.) What does the `Runner` do with this environment.
+
+### Configuration
+
+If you check out the constructor of `Mutant::CLI` you'll see this:
+
+```Ruby
+def initialize(arguments)
+  @config = Config::DEFAULT
+
+  parse(arguments)
+end
+```
+
+Ok, nice and easy. First we'll set the configuration to our defaults, then we're parsing the cli arguments and overwrite our configuration based on this.
+
+What does the default configuration look like?
+For that you'll have to check out the top level library file, so lib/mutant.rb:
```

Comment:
For reference, this is the only location in mutant I re-open a namespace. The whole require sequence is cycle free (and thus loads much faster than for example with autloads + backtracing like rails does).

Maintaining the require list explicitly and having "one" well choosen cycle is far more maintenance friendly deterministic and performant (assuming a full! load) than other solutions.